tests/__tests__/linear/drain.ava.ts

```diff
@@ -0,0 +1,322 @@
+import { NearAccount, NEAR, Gas } from "near-workspaces-ava";
+import { assertFailure, initWorkSpace, createStakingPool } from "./helper";
+
+const workspace = initWorkSpace();
+
+async function setManager(
+  owner: NearAccount,
+  contract: NearAccount,
+  manager: NearAccount
+) {
+  await owner.call(
+    contract,
+    'add_manager',
+    {
+      new_manager_id: manager.accountId
+    }
+  );
+}
+
+function assertValidatorAmountHelper (
+  test: any,
+  contract: NearAccount,
+  owner: NearAccount
+) {
+  return async function (
+    validator: NearAccount, 
+    stakedAmount: string,
+    unstakedAmount: string
+  ) {
+    // 1. make sure validator has correct balance
+    test.is(
+      await validator.view('get_account_staked_balance', { account_id: contract.accountId }),
+      NEAR.parse(stakedAmount).toString()
+    );
+    test.is(
+      await validator.view('get_account_unstaked_balance', { account_id: contract.accountId }),
+      NEAR.parse(unstakedAmount).toString()
+    );
+
+    // 2. make sure contract validator object is synced
+    const v: any = await owner.call(
+      contract,
+      'get_validator',
+      {
+        validator_id: validator.accountId
+      }
+    );
+    const staked = NEAR.from(v.staked_amount);
+    const unstaked = NEAR.from(v.unstaked_amount);
+    test.is(
+      staked.toString(),
+      NEAR.parse(stakedAmount).toString()
+    );
+    test.is(
+      unstaked.toString(),
+      NEAR.parse(unstakedAmount).toString()
+    );
+  }
+}
+
+async function stakeAll (manager: NearAccount, contract: NearAccount) {
+  let run = true;
+  while (run) {
+    run = await manager.call(
+      contract,
+      'epoch_stake',
+      {},
+      {
+        gas: Gas.parse('200 Tgas')
+      }
+    );
+  }
+}
+
+workspace.test('Non-manager call drain methods', async (test, {contract, alice}) => {
+    await assertFailure(
+        test,
+        alice.call(
+            contract,
+            'drain_unstake',
+            {
+                validator_id: 'foo'
+            }
+        ),
+        'Only manager can perform this action'
+    );
+
+    await assertFailure(
+        test,
+        alice.call(
+            contract,
+            'drain_withdraw',
+            {
+                validator_id: 'foo'
+            }
+        ),
+        'Only manager can perform this action'
+    );
+});
+
+workspace.test('drain constraints', async (test, {contract, root, owner, alice}) => {
+  const manager = alice;
+  await setManager(owner, contract, manager);
+
+  const v1 = await createStakingPool(root, 'v1');
+  // add validator
+  await manager.call(
+    contract,
+    'add_validator',
+    {
+      validator_id: v1.accountId,
+      weight: 10
+    }
+  );
+
+  // user stake
+  await alice.call(
+    contract,
+    'deposit_and_stake',
+    {},
+    {
+      attachedDeposit: NEAR.parse('50')
+    }
+  );
+
+  // run stake
+  await manager.call(
```

Comment:
maybe we can make the call with another executor / operator such as `bob`, which requires no `manager` role. just to make it clear that the `drain_*` functions require `manager` permission, but `epoch_*` actions are not.

Reply:
good point, will do